packages/browser/src/profiling/integration.ts

```diff
@@ -15,34 +23,67 @@ import { addProfilingExtensionMethods } from './hubextensions';
  */
 export class BrowserProfilingIntegration implements Integration {
   public readonly name: string = 'BrowserProfilingIntegration';
+  public getCurrentHub?: () => Hub = undefined;
 
   /**
    * @inheritDoc
    */
-  public setupOnce(addGlobalEventProcessor: (callback: EventProcessor) => void): void {
-    // Patching the hub to add the extension methods.
-    // Warning: we have an implicit dependency on import order and we will fail patching if the constructor of
-    // BrowserProfilingIntegration is called before @sentry/tracing is imported. This is because we need to patch
-    // the methods of @sentry/tracing which are patched as a side effect of importing @sentry/tracing.
-    addProfilingExtensionMethods();
-
-    // Add our event processor
-    addGlobalEventProcessor(this.handleGlobalEvent.bind(this));
-  }
+  public setupOnce(addGlobalEventProcessor: (callback: EventProcessor) => void, getCurrentHub: () => Hub): void {
+    this.getCurrentHub = getCurrentHub;
+    const client = this.getCurrentHub().getClient() as BrowserClient;
 
-  /**
-   * @inheritDoc
-   */
-  public handleGlobalEvent(event: Event): Event {
-    const profileId = event.contexts && event.contexts['profile'] && event.contexts['profile']['profile_id'];
-
-    if (profileId && typeof profileId === 'string') {
-      if (__DEBUG_BUILD__) {
-        logger.log('[Profiling] Profiling event found, caching it.');
-      }
-      PROFILING_EVENT_CACHE.add(profileId, event);
-    }
+    if (client && typeof client.on === 'function') {
+      client.on('startTransaction', (transaction: Transaction) => {
+        wrapTransactionWithProfiling(transaction);
+      });
+
+      client.on('beforeEnvelope', (envelope): void => {
+        // if not profiles are in queue, there is nothing to add to the envelope.
+        if (!PROFILE_MAP.size) {
+          return;
+        }
+
+        const profiledTransactionEvents = findProfiledTransactionsFromEnvelope(envelope);
+        if (!profiledTransactionEvents.length) {
+          return;
+        }
 
-    return event;
+        const profilesToAddToEnvelope: Profile[] = [];
+
+        for (const profiledTransaction of profiledTransactionEvents) {
+          const profile_id =
+            profiledTransaction &&
+            profiledTransaction.contexts &&
+            profiledTransaction.contexts['profile'] &&
+            profiledTransaction.contexts['profile']['profile_id'] as string
```

Comment:
Nit: you might be able to save a few key strokes if you got the profile context first, then get the profile id instead of the profile id then the profile context like you have here.